### PR TITLE
Use *cli.Context for configuration

### PR DESCRIPTION
**Pull request by zpatrick**
**What does this pull request do?**
* Replaces the current configuration pattern(s) to use only `cli.Flag` objects within the `config` package, and use `*cli.Context` objects for callers. 
* Solidifies relationship between the layer0 terraform module outputs and their corresponding flag in the `config` package (terraform output == flag name)
* Updates the layer0 module's `Dockerrun.aws.json` environment variables to match their corresponding values in `config`. 
* Moves the `TestContext` helper(s) in `cli/command` and `setup/command` into the `config` package. 

**How should this be tested?**
* Unit Tests
* Using `l0-setup -d endpoint` to populate envvars, run layer0 api locally 
* Run layer0 api in aws using `flow api`

**Checklist**
- [ ] Unit tests
- ~[ ] Smoke tests (if applicable)~
- ~[ ] System tests (if applicable)~
- ~[ ] Documentation (if applicable)~

### CODE REVIEW

File: api/provider/aws/deploy_create_test.go [hunk 1]
```diff
@@ -51,19 +52,24 @@ func TestDeploy_createTags(t *testing.T) {
...
+		config.FlagAWSLogGroup.GetName(): "test_group",
+		config.FlagAWSRegion.GetName():   "test_region",
+	})
+
+	deploy := NewDeployProvider(nil, nil, c)
 
 	container := &ecs.ContainerDefinition{}
 	container.SetName("test_name")
-	logConfig := &ecs.LogConfiguration{
+	logContext := &ecs.LogConfiguration{
```
Comment (zpatrick) [inline api/provider/aws/deploy_create_test.go]: need to fix: bad `sed` command

File: api/provider/aws/environment_create.go [hunk 2]
```diff
@@ -81,13 +83,15 @@ func (e *EnvironmentProvider) Create(req models.CreateEnvironmentRequest) (strin
 		return "", err
 	}
 
-	launchConfigName := fqEnvironmentID
+	launchContextName := fqEnvironmentID
+	instanceProfile := e.Context.String(config.FlagAWSInstanceProfile.GetName())
+	sshKeyPair := e.Context.String(config.FlagAWSSSHKey.GetName())
 	if err := e.createLC(
-		launchConfigName,
+		launchContextName,
```
Comment (zpatrick) [inline api/provider/aws/environment_create.go]: ditto

File: api/provider/aws/environment_create.go [hunk 3]
```diff
@@ -99,12 +103,13 @@ func (e *EnvironmentProvider) Create(req models.CreateEnvironmentRequest) (strin
 	}
 
 	autoScalingGroupName := fqEnvironmentID
+	privateSubnets := e.Context.StringSlice(config.FlagAWSPrivateSubnets.GetName())
 	if err := e.createASG(
 		autoScalingGroupName,
-		launchConfigName,
+		launchContextName,
```
Comment (zpatrick) [inline api/provider/aws/environment_create.go]: ditto

File: api/provider/aws/environment_create.go [hunk 4]
```diff
@@ -141,7 +146,7 @@ func (e *EnvironmentProvider) authorizeSGSelfIngress(groupID string) error {
 }
 
 func (e *EnvironmentProvider) createLC(
-	launchConfigName string,
+	launchContextName string,
```
Comment (zpatrick) [inline api/provider/aws/environment_create.go]: ditto

File: api/provider/aws/environment_create.go [hunk 5]
```diff
@@ -150,7 +155,7 @@ func (e *EnvironmentProvider) createLC(
 	userData string,
 ) error {
 	input := &autoscaling.CreateLaunchConfigurationInput{}
-	input.SetLaunchConfigurationName(launchConfigName)
+	input.SetLaunchConfigurationName(launchContextName)
```
Comment (zpatrick) [inline api/provider/aws/environment_create.go]: ditto

File: api/provider/aws/environment_create.go [hunk 6]
```diff
@@ -171,7 +176,7 @@ func (e *EnvironmentProvider) createLC(
 
 func (e *EnvironmentProvider) createASG(
 	autoScalingGroupName string,
-	launchConfigName string,
+	launchContextName string,
```
Comment (zpatrick) [inline api/provider/aws/environment_create.go]: ditto

File: api/provider/aws/environment_create.go [hunk 7]
```diff
@@ -185,7 +190,7 @@ func (e *EnvironmentProvider) createASG(
 
 	input := &autoscaling.CreateAutoScalingGroupInput{}
 	input.SetAutoScalingGroupName(autoScalingGroupName)
-	input.SetLaunchConfigurationName(launchConfigName)
+	input.SetLaunchConfigurationName(launchContextName)
```
Comment (zpatrick) [inline api/provider/aws/environment_create.go]: ditto

File: api/provider/aws/environment_delete.go [hunk 8]
```diff
@@ -16,15 +16,15 @@ import (
...
+	fqEnvironmentID := addLayer0Prefix(e.Context, environmentID)
 
 	autoScalingGroupName := fqEnvironmentID
 	if err := e.deleteASG(autoScalingGroupName); err != nil {
 		return err
 	}
 
-	launchConfigName := fqEnvironmentID
-	if err := e.deleteLC(launchConfigName); err != nil {
+	launchContextName := fqEnvironmentID
```
Comment (zpatrick) [inline api/provider/aws/environment_delete.go]: ditto

File: api/provider/aws/environment_delete.go [hunk 9]
```diff
@@ -73,9 +73,9 @@ func (e *EnvironmentProvider) deleteASG(autoScalingGroupName string) error {
 	return nil
 }
 
-func (e *EnvironmentProvider) deleteLC(launchConfigName string) error {
+func (e *EnvironmentProvider) deleteLC(launchContextName string) error {
 	input := &autoscaling.DeleteLaunchConfigurationInput{}
-	input.SetLaunchConfigurationName(launchConfigName)
+	input.SetLaunchConfigurationName(launchContextName)
```
Comment (zpatrick) [inline api/provider/aws/environment_delete.go]: ditto

File: api/provider/aws/environment_read.go [hunk 10]
```diff
@@ -30,8 +30,8 @@ func (e *EnvironmentProvider) Read(environmentID string) (*models.Environment, e
 		return nil, err
 	}
 
-	launchConfigName := aws.StringValue(autoScalingGroup.LaunchConfigurationName)
-	launchConfig, err := e.readLC(launchConfigName)
+	launchContextName := aws.StringValue(autoScalingGroup.LaunchConfigurationName)
```
Comment (zpatrick) [inline api/provider/aws/environment_read.go]: ditto

File: api/provider/aws/environment_read.go [hunk 11]
```diff
@@ -44,29 +44,29 @@ func (e *EnvironmentProvider) Read(environmentID string) (*models.Environment, e
 	model.MinScale = int(aws.Int64Value(autoScalingGroup.MinSize))
 	model.CurrentScale = int(aws.Int64Value(autoScalingGroup.DesiredCapacity))
 	model.MaxScale = int(aws.Int64Value(autoScalingGroup.MaxSize))
-	model.InstanceType = aws.StringValue(launchConfig.InstanceType)
+	model.InstanceType = aws.StringValue(launchContext.InstanceType)
```
Comment (zpatrick) [inline api/provider/aws/environment_read.go]: ditto

File: api/provider/aws/environment_read.go [hunk 12]
```diff
@@ -44,29 +44,29 @@ func (e *EnvironmentProvider) Read(environmentID string) (*models.Environment, e
...
 	model.SecurityGroupID = aws.StringValue(securityGroup.GroupId)
-	model.AMIID = aws.StringValue(launchConfig.ImageId)
+	model.AMIID = aws.StringValue(launchContext.ImageId)
 
 	return model, nil
 }
 
-func (e *EnvironmentProvider) readLC(launchConfigName string) (*autoscaling.LaunchConfiguration, error) {
+func (e *EnvironmentProvider) readLC(launchContextName string) (*autoscaling.LaunchConfiguration, error) {
```
Comment (zpatrick) [inline api/provider/aws/environment_read.go]: ditto

File: api/provider/aws/environment_read.go [hunk 13]
```diff
@@ -44,29 +44,29 @@ func (e *EnvironmentProvider) Read(environmentID string) (*models.Environment, e
...
+	input.SetLaunchConfigurationNames([]*string{aws.String(launchContextName)})
 
 	output, err := e.AWS.AutoScaling.DescribeLaunchConfigurations(input)
 	if err != nil {
 		return nil, err
 	}
 
 	for _, lc := range output.LaunchConfigurations {
-		if aws.StringValue(lc.LaunchConfigurationName) == launchConfigName {
+		if aws.StringValue(lc.LaunchConfigurationName) == launchContextName {
```
Comment (zpatrick) [inline api/provider/aws/environment_read.go]: ditto

File: common/config/setup.go [hunk 14]
```diff
@@ -8,9 +8,19 @@ import (
...
 	return []cli.Flag{
-		cli.BoolFlag{
-			Name:   fmt.Sprintf("d, %s", FLAG_DEBUG),
-			EnvVar: ENVVAR_DEBUG,
-		},
+		FlagDebug,
 	}
 }
+
+func ValidateSetupContext(c *cli.Context) error {
```
Comment (zpatrick) [inline common/config/setup.go]: I should have a helper function that all of these `Validate` functions use

File: setup/module/api/Dockerrun.aws.json [hunk 15]
```diff
@@ -19,24 +19,25 @@
...
-            { "name": "LAYER0_AWS_DYNAMO_TAG_TABLE", "value": "${dynamo_tag_table}" },
-            { "name": "LAYER0_AWS_DYNAMO_JOB_TABLE", "value": "${dynamo_job_table}" },
-            { "name": "LAYER0_AWS_DYNAMO_LOCK_TABLE", "value": "${dynamo_lock_table}" },
-	    { "name": "LAYER0_AUTH_TOKEN", "value": "${api_auth_token}" },
+            { "name": "LAYER0_AWS_LINUX_AMI", "value": "${linux_service_ami}" },
+            { "name": "LAYER0_AWS_WINDOWS_AMI", "value": "${windows_service_ami}" },
+            { "name": "LAYER0_AWS_TAG_TABLE", "value": "${dynamo_tag_table}" },
+            { "name": "LAYER0_AWS_JOB_TABLE", "value": "${dynamo_job_table}" },
+            { "name": "LAYER0_AWS_LOCK_TABLE", "value": "${dynamo_lock_table}" },
+	    { "name": "LAYER0_TOKEN", "value": "${api_auth_token}" },
```
Comment (zpatrick) [inline setup/module/api/Dockerrun.aws.json]: [nit] I should fix this formatting

File: api/controllers/admin.go [hunk 16]
```diff
@@ -37,11 +38,11 @@ func (a *AdminController) Routes() []*fireball.Route {
 
 func (a *AdminController) GetConfig(c *fireball.Context) (fireball.Response, error) {
 	model := models.APIConfig{
-		Instance:       a.Config.Instance(),
-		VPCID:          a.Config.VPC(),
+		Instance:       a.Context.String(config.FlagInstance.GetName()),
```
Comment (diemonster) [inline api/controllers/admin.go]: Is there any way we could simplify getting these values?

File: api/main.go [hunk 17]
```diff
@@ -49,40 +49,43 @@ func main() {
...
+		delay := c.Duration(config.FlagAWSRequestDelay.GetName())
 		ticker := time.Tick(delay)
 		session.Handlers.Send.PushBack(func(r *request.Request) {
 			<-ticker
 		})
 
 		client := awsclient.NewClient(session)
-		tagStore := tag.NewDynamoStore(session, cfg.DynamoTagTable())
-		jobStore := job.NewDynamoStore(session, cfg.DynamoJobTable())
+		tagStore := tag.NewDynamoStore(session, c.String(config.FlagAWSTagTable.GetName()))
```
Comment (diemonster) [inline api/main.go]: similar to my last comment re: simplifying getting such values, but it seems like having to know that `config.FlagAWSTagTable.GetName()` must be converted by `c.String()` is awkward

File: setup/module/api/load_balancer.tf [hunk 18]
```diff
@@ -65,7 +65,7 @@ resource "aws_elb" "api" {
     healthy_threshold   = 2
     unhealthy_threshold = 2
     timeout             = 5
-    target              = "HTTP:80/admin/health"
```
Comment (diemonster) [inline setup/module/api/load_balancer.tf]: Why is this being changed?

File: api/controllers/admin.go (hunk 16 above)
Comment (zpatrick) [reply api/controllers/admin.go]: Nothing that i can can think of that would require either:
* An interface, which would undo the main point of this pr
* Helper functions that would be more-or-less redundant, e.g. 
```go
foo := config.String(config.FlagInstance)
```

File: api/main.go (hunk 17 above)
Comment (zpatrick) [reply api/main.go]: see above

File: setup/module/api/load_balancer.tf (hunk 18 above)
Comment (zpatrick) [reply setup/module/api/load_balancer.tf]: The `admin/health` endpoint requires auth. It's easier to change this than to have one-off routes in `api`

File: common/config/flags.go [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+var (
+	FlagDebug = cli.BoolFlag{
+		Name:   "debug",
+		EnvVar: "LAYER0_DEBUG",
+		Usage:  "show debug output",
+	}
+	FlagEndpoint = cli.StringFlag{
+		Name:   "endpoint",
+		Value:  DefaultEndpoint,
+		EnvVar: "LAYER0_ENDPOINT",
```
Comment (jparsons04) [inline common/config/flags.go]: Something I'm concerned about is the renaming of the environment variables themselves. This used to be called `LAYER0_API_ENDPOINT`. I'm concerned that by doing this, we are rendering older api image versions obsolete, because users would be forced to upgrade and change their environment variable structure. Other environment variables further below are also name changes.

File: common/config/flags.go [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+		Usage:  "endpoint of the Layer0 API",
+	}
+	FlagToken = cli.StringFlag{
+		Name:   "token",
+		EnvVar: "LAYER0_TOKEN",
```
Comment (jparsons04) [inline common/config/flags.go]: This was `LAYER0_AUTH_TOKEN`

File: common/config/flags.go [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+		Usage:  "auth token of the Layer0 API",
+	}
+	FlagAWSAccessKey = cli.StringFlag{
+		Name:   "aws-access-key",
+		EnvVar: "LAYER0_AWS_ACCESS_KEY",
```
Comment (jparsons04) [inline common/config/flags.go]: This was `LAYER0_AWS_ACCESS_KEY_ID`

File: common/config/flags.go [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+		Usage:  "access key portion of an AWS key",
+	}
+	FlagAWSSecretKey = cli.StringFlag{
+		Name:   "aws-secret-key",
+		EnvVar: "LAYER0_AWS_SECRET_KEY",
```
Comment (jparsons04) [inline common/config/flags.go]: This was `LAYER0_AWS_SECRET_ACCESS_KEY`

File: setup/module/variables.tf [hunk 23]
```diff
@@ -16,8 +16,8 @@ variable "username" {}
 
 variable "password" {}
 
-variable "time_between_requests" {
-  default = "10ms"
+variable "request_delay" {
+  default = "1ms"
```
Comment (sesh-kebab) [inline setup/module/variables.tf]: Is this change intentional?

File: api/controllers/admin.go (hunk 16 above)
Comment (sesh-kebab) [reply api/controllers/admin.go]: I'm not a 100% sure here either. After an initial look:
* I will need more time to review this
* The trade-off for simplifying the config package is very apparent

File: common/config/flags.go (hunk 19 above)
Comment (zpatrick) [reply common/config/flags.go]: If we weren't in the process of a new minor version I would agree with you, but this refactor is already a minor version bump. Older versions of the API, CLI, and Setup binaries are already incompatible with this version in a number of different ways. 

API images are tightly coupled with their respective version of `l0-setup`. The upgrade/changes to their environment variable structure is done for them via the `l0-setup`, which handled with `l0-setup init` or `l0-setup upgrade`. 

So, within minor version updates I agree we should avoid changing these variables, but I see no problem with doing it between minor version updates.

File: setup/module/variables.tf (hunk 23 above)
Comment (zpatrick) [reply setup/module/variables.tf]: yeah, didn't see any reason for it to be relatively slow by default.

File: common/config/flags.go (hunk 19 above)
Comment (diemonster) [reply common/config/flags.go]: This is a good comment to think about if only because we're going to have more breaking changes in this release than we have had in a very long time. Historically, this has caused us headaches for people that don't upgrade across the board, so something to keep in mind when we update docs, etc.

File: setup/module/variables.tf (hunk 23 above)
Comment (zpatrick) [reply setup/module/variables.tf]: (wouldn't be opposed to using nanoseconds)

**Comment by jparsons04** [comment]
You are missing the `config.APIConfig` -> `*cli.Context` changes in `provider/aws/admin.go`

File: api/main.go [hunk 24]
```diff
@@ -48,37 +48,44 @@ func main() {
...
-		adminProvider := aws.NewAdminProvider(client, tagStore, cfg)
-		deployProvider := aws.NewDeployProvider(client, tagStore, cfg)
-		environmentProvider := aws.NewEnvironmentProvider(client, tagStore, cfg)
-		loadBalancerProvider := aws.NewLoadBalancerProvider(client, tagStore, cfg)
-		serviceProvider := aws.NewServiceProvider(client, tagStore, cfg)
-		taskProvider := aws.NewTaskProvider(client, tagStore, cfg)
+		tagStore := tag.NewDynamoStore(session, c.String(config.FlagAWSTagTable.GetName()))
+		jobStore := job.NewDynamoStore(session, c.String(config.FlagAWSJobTable.GetName()))
+
+		if err := addAPIEntityTags(tagStore); err != nil {
```
Comment (jparsons04) [inline api/main.go]: This appears to be the only reference to `addAPIEntityTags()` and to me looks to be undefined elsewhere in the set of changes you've made.

File: setup/module/api/Dockerrun.aws.json [hunk 25]
```diff
@@ -19,24 +19,25 @@
...
+            { "name": "LAYER0_AWS_ACCESS_KEY", "value": "${access_key}" },
+            { "name": "LAYER0_AWS_SECRET_KEY", "value": "${secret_key}" },
             { "name": "LAYER0_AWS_REGION", "value": "${region}" },
             { "name": "LAYER0_AWS_PUBLIC_SUBNETS", "value": "${public_subnets}" },
             { "name": "LAYER0_AWS_PRIVATE_SUBNETS", "value": "${private_subnets}" },
-            { "name": "LAYER0_AWS_ECS_INSTANCE_PROFILE", "value": "${ecs_instance_profile}" },
-            { "name": "LAYER0_AWS_VPC_ID", "value": "${vpc_id}" },
+            { "name": "LAYER0_AWS_INSTANCE_PROFILE", "value": "${ecs_instance_profile}" },
+            { "name": "LAYER0_AWS_VPC", "value": "${vpc_id}" },
+            { "name": "LAYER0_AWS_SSH_KEY", "value": "${ssh_key_pair}" },
```
Comment (jparsons04) [inline setup/module/api/Dockerrun.aws.json]: I opened issue #455 to address this specific change.

File: setup/module/api/load_balancer.tf (hunk 18 above)
Comment (jparsons04) [reply setup/module/api/load_balancer.tf]: I opened issue #456 to address this change.

File: api/main.go (hunk 24 above)
Comment (zpatrick) [reply api/main.go]: bad merge on my part. fixed.

**Comment by zpatrick** [comment]
closing for now to reduce clutter
